_posts/2023-11-17-transactional-outbox.md

```diff
@@ -0,0 +1,295 @@
+3. **Observing Event Handling:** As you create users through the endpoint, keep an eye on the logs of the RabbitMqConsumer. 
+You should see the `UserCreatedEvent` payloads being logged, indicating that the events are successfully being consumed from RabbitMQ.
+
+For convenience, the repository includes a `sample-requests.http` file,
+which provides easy access to all the endpoints if you are using IntelliJ IDEA.
+This file simplifies the process of making requests to the application for testing purposes.
+
+## Explore the Code
+
+The `UserController` controller provides an endpoint to create users.
+```kotlin
+@PostMapping("/users")
+fun createUser(@RequestBody request: UserCreateRequestDto): User {
+    return userService.createUser(request.name)
+}
+```
+
+When a user is created, a `UserCreatedEvent` is published. 
+This is done using the OutboxPublisher to persist the event in the outbox first, and then forward it to RabbitMQ.
+```kotlin
+class UserService(
+    private val userRepository: UserRepository, 
+    private val eventPublisher: EventPublisher) {
+    
+    fun createUser(name: String): User {
+        return userRepository.save(User(UUID.randomUUID(), name)).also {
+            eventPublisher.publish(UserCreatedEvent(it.id, it.name))
+        }
+    }
+}
+
+class OutboxPublisher(private val outbox: TransactionOutbox): EventPublisher {
+    override fun publish(event: DomainEvent) {
+        outbox.schedule(RabbitMqSender::class.java).send(event)
+    }
+}
+```
+
+The `schedule` method leads to the persisting of the task, which involves sending the event to RabbitMQ. 
+This task is stored in the outbox. 
+Then, it schedules the execution of this task for after the current transaction has been committed. 
+This ensures that the event is only sent after the transaction is successfully completed.
+
+In the context of handling potential failures during event sending after a transaction, the `TransactionalOutboxScheduler` class provides a robust fallback mechanism. 
+This class periodically executes a run method that continuously calls `transactionOutbox.flush()`.
+The `flush` method in `TransactionOutbox` is responsible for finding and retrying events that haven't been successfully executed after a certain amount of time. 
+It specifically targets events that were initially set up to be executed but haven't been completed as expected and have not exceeded a predefined number of retry attempts.
+
+```kotlin
+class TransactionalOutboxScheduler(private val transactionOutbox: TransactionOutbox) {
+    @Scheduled(fixedDelay = 2000)
+    fun run() {
+        while (transactionOutbox.flush()) {
+            // NOP
+        }
+    }
+}
+```
+
+To simulate errors in event sending, utilize the `/failing-events` endpoint. 
+This endpoint generates a `FailingEvent` and attempts to send it to RabbitMQ. 
+The `RabbitMqSender` is programmed to mimic an error by intentionally failing once for each instance of `FailingEvent`.
+
```

Comment:
you could make this ending smoother by adding another sentence such as: "Play around and have fun exploring the outbox pattern!"

Reply:
maybe add a little challenge for the reader? if they downloaded the repo, what else could they try to add on their own? without explaining the solution. "tell us in the comments how you solved this challenge"?